### ISSUE TITLE
Switch pre-commit hooks to ruff (100x faster pre-commit, updated code style)

### ISSUE THREAD

**Issue by MattHag** [pwr-Solaar/Solaar#2295]
**Information**
- Solaar version: 1.1.11.rc4

**Is your feature request related to a problem? Please describe.**
Unique formatting of code and linting is enforced by CI checks using a pre-commit hook with _yapf_, _isort_ and _flake8_. 

Especially, the yapf code formatter and flake8 linter are slow compared to modern tools. This renders yapf 

- useless as auto-formatter during development, 
- slows down pre-commits and 
- slows down the format and linter checks in the CI pipeline. 

**Describe the solution you'd like**
The fastest Python code formatter on the market is currently [ruff format](https://docs.astral.sh/ruff/formatter/), which is 100x faster than _yapf_. In addition to code formatting, the [ruff linter can also replace _isort_ and _flake8_](https://docs.astral.sh/ruff/linter/). That would reduce the development dependencies from three to one and dramatically speed up pre-commit hooks.

- Upgrade the pre-commit hooks to speed them up. 
- Introduce a formatter, that's fast enough to use during development, e.g. by running `pre-commit run -a` or just ruff format. You then leverage a shortcut for auto-formatting or your IDE's format-on-save feature, ruff plugin etc. This saves you from reformatting, lists, dictionaries etc. during coding manually.

**Describe alternatives you've considered**
The Black formatter is roughly twice as fast as yapf, but still slow in comparison to ruff. The capabilities of [ruff are now 99.9% on par with black, but it runs 30x faster](https://astral.sh/blog/the-ruff-formatter). That's a big difference as pre-commit hook, but also results in a faster CI pipeline.

**Comment by pfps** [comment]
Sounds like a good idea for after 1.1.11.

**Comment by pfps** [comment]
@MattHag Go ahead and create a PR for this.

**Comment by pfps** [comment]
@MattHag One potential issue is that it currently doesn' t look as if roff has the --fix flag when run from pre-commit.   That would be useful.

**Comment by pfps** [comment]
Another potential issue is that the line-length in Solaar is currently set to 127, not 120.

**Comment by MattHag** [comment]
That's all not an issue, might fix it later today.

**Comment by pfps** [comment]
@MattHag 
When I commit locally, ruff doesn't change the files to fix formatting issues.  How can I get it to do that?

**Comment by MattHag** [comment]
Install latest pre-commit hooks
`pre-commit install`
Run all pre-commit hooks on command
`pre-commit run -a`

Run ruff lint with autofix
`make lint`
Run ruff autoformat
`make format`

**Comment by pfps** [comment]
When I do a commit ruff doesn't fix the files, like yapf and isort used to.

For example,

```
idefix Solaar> git add lib/logitech_receiver/settings_templates.py 
idefix Solaar> git commit --amend --no-edit
check python ast.........................................................Passed
check builtin type constructor use.......................................Passed
check for merge conflicts................................................Passed
check yaml...........................................(no files to check)Skipped
check toml...........................................(no files to check)Skipped
debug statements (python)................................................Passed
fix end of files.........................................................Passed
trim trailing whitespace.................................................Passed
ruff lint................................................................Passed
ruff format..............................................................Failed
- hook id: ruff-format
- exit code: 1

--- lib/logitech_receiver/settings_templates.py
+++ lib/logitech_receiver/settings_templates.py
@@ -1504,8 +1504,8 @@
 # a field for the battery information - eventually to be split into the various kinds of batteries
 class Battery(_Field):
     name = "battery"
-    label = _('Battery')
-    description = _('Battery information read from device')
+    label = _("Battery")
+    description = _("Battery information read from device")
     feature = True  # need to see how to instantiate for both register and feature devices
     register = True
 

1 file would be reformatted
```

But the file is not changed.

**Comment by MattHag** [comment]
Ah, wait, I switched the pre-commit to `ruff format --diff` as nice check for GitHub CI. But let's try with the default format but return non-zero.

**Comment by MattHag** [comment]
Try again

**Comment by pfps** [comment]
That's got it, thanks.

**Comment by pfps** [comment]
Fixed by #2316
